elastic/beats
# [Ingest Manager] Retryable downloads of beats

## What does this PR do?

Background:
when agent downloads an artifact and checksum does not match it yields a failure, but then it might occur that when download is performed again due to new config or whatever, download is skipped (because download was successful for some reason or packed artifacts are invalid). 
Agent cleans up downloaded artifact only in case download yields error. so if this does not yield error but artifact is corrupted we might end up in a loop because it will try to verify artifact it find out it's incorrect and continues with failure... and so on

This PR changes this behavior a bit.

In case Verify fails. it cleans up downloaded artifacts (artifact + hash).

It also introduces retryable block within operation flow.
In this case we know than=t download+verify might be error prone so we can retry them if failure happens. (only if retry.enabled == true) 

What this means for agent is that when it tries to install from corrupted artifact, it will remove artifact during Verify and re-download it again.

## Why is it important?

Make download scneario more robust and repair loop faster

## Checklist


## How to test

- Build a snapshot package
- Modify one of sha files
- enable retry by setting `retry.enabled: true`
- run agent

See it fails with packed artifact, waits 30s and then downloads artifact from web
